pkg/shortnames/errors.go

```diff
@@ -0,0 +1,118 @@
+package shortnames
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/containers/image/v5/pkg/sysregistriesv2"
+	"github.com/containers/image/v5/types"
+	"github.com/pkg/errors"
+)
+
+// GuessPullErrors returns an error that _guesses_ why the specified short
+// name could not resolve to a successful pull candidate along with the
+// specified pull errors.  This can be helpful for giving some hints to users
+// to resolve pull errors.  Note that nil is returned if it's not a short name.
+//
+// GuessPullErrors is expected to be called if none of the `Resolved`
+// candidates from `Resolve` lead to a successfully pulled image.  The returned
+// error should be displayed to users as is and not be wrapped any further.
+func GuessPullError(ctx *types.SystemContext, name string, pullErrors []error) error {
```

Comment:
Resolve does not attempt to pull. Both Buildah and Podman tried to reason over why a pull may have failed and that's attempt to consolidate the code and give the same errors.

I spent some time thinking of embedding the logic into `Resolved` and `Resolve` but I found that would over-complicate things. In the end, it's a way to have a consistent formatting plus some hints to resolve the errors as Podman and Buildah did before.

Reply:
- The values returned currently don’t _at all_ guess “why a pull may have failed”. That might be very useful to build a heuristic for, but the current implementation is not trying. (I guess, if that’s the real goal, build a placeholder heuristic that fails to make a recommendation (`if true { }`?), just to show where it would go and what the output would look like.)
- The thing the implementation _is_ doing, providing context for why a particular set of qualified names was chosen, is already turning out to be difficult to get right.

Reply:
Compare also the history of Podman’s `pullGoal.{usedSearchRegistries,searchedRegistries}`: That was added because the previous version was also guessing what probably happened only after a failure, and was guessing wrong and/or getting out of sync with the actual decision logic.

Reply:
I guess you're right. We're late, so I am shielding bigger changes but this one makes sense to push through.

How about we:
* Record during `Resolve` how a given `Resolved` came to be (alias, USR, none)
* Add a new method `(*Resolved) WrapError(err error) error` which will add the context information to the error. This way each error will give us more information.

Future cookies/work: record during config loading where the USRs come from and where each alias comes from.

Reply:
How would the caller know what single `Resolved` to use? It can’t well iterate though all the `Resolved` candidate, call `WrapPullError` on all of them, and combine the result, because that would usually repeat the same explanation N times.

alias/USR/none is a single decision that feels like a single return value to me (see `ResolutionResult.Rationale` below, OTOH that has unresolved issues as to how the record/SystemContext values participate in recording outcomes).

Admittedly if (some variant of) `Resolve` starts adding `localhost/` entries, we _may_ need different explanation for entries (e.g. “`localhost/` comes from application policy, the others come from USR”).

Or maybe combine both, and turn `GuessPullError([]error)` into `ResolutionResult.WrapError(errors []error)` , requiring that `len(errors) == len(ResolutionResult.Resolved)` but otherwise giving us the option to record any amount of data that later turns out to be useful as private members of `ResolutionResult`.

(As usual, my naming suggestions suck. Please improve on them.)

Reply:
> How would the caller know what single Resolved to use? It can’t well iterate though all the Resolved candidate, call WrapPullError on all of them, and combine the result, because that would usually repeat the same explanation N times.

Consider:
```golang
candidates, _ := shortnames.Resolve(...)

for _, c := range candidates {
   _, err := pull(c)
   if err != nil {
     err = c.WrapPullError(err)
   }
}
```
AFAICS that would eliminate the need for `GuessPullErrors` entirely.

Reply:
At least that would explain where a given fully-qualified reference originated from.

Reply:
That creates one `err` value per loop iteration; and without further heuristics the caller will have to show all of them (the way `GuessPullError` did). So all of them will include exactly the same rationale (especially in the case of USR), the and the caller will end up with ~2-5 errors all starting with the same “%q has been completed with an unqualified-search registry” prefix.

True, it would also show the fully-qualified name, but IIRC that shows up in error messages anyway (and it’s also the part that the caller of `Resolve` can easily do itself, if truly necessary).